File: long_questions/DSA/Leetcode/golang/10_Tree_DFS/104_Maximum_Depth_of_Binary_Tree.go
package main

import "fmt"

// Definition for a binary tree node.
type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

// 104. Maximum Depth of Binary Tree
// Time: O(N), Space: O(H) where H is tree height (recursion stack)
func maxDepth(root *TreeNode) int {
	if root == nil {
		return 0
	}
	
	leftDepth := maxDepth(root.Left)
	rightDepth := maxDepth(root.Right)
	
	return max(leftDepth, rightDepth) + 1
}

// Iterative BFS approach
func maxDepthBFS(root *TreeNode) int {
	if root == nil {
		return 0
	}
	
	queue := []*TreeNode{root}
	depth := 0
	
	for len(queue) > 0 {
		levelSize := len(queue)
		depth++
		
		for i := 0; i < levelSize; i++ {
			node := queue[0]
			queue = queue[1:]
			
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
	}
	
	return depth
}

// Helper function to create a binary tree from slice
func createTree(nums []interface{}) *TreeNode {
	if len(nums) == 0 {
		return nil
	}
	
	nodes := make([]*TreeNode, len(nums))
	for i, val := range nums {
		if val == nil {
			nodes[i] = nil
		} else {
			nodes[i] = &TreeNode{Val: val.(int)}
		}
	}
	
	for i := 0; i < len(nums); i++ {
		left := 2*i + 1
		right := 2*i + 2
		
		if left < len(nums) {
			nodes[i].Left = nodes[left]
		}
		if right < len(nums) {
			nodes[i].Right = nodes[right]
		}
	}
	
	return nodes[0]
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Tree Depth-First Search (DFS)
- **Recursive Traversal**: Visit each node exactly once
- **Base Case**: Return 0 for nil nodes
- **Recursive Case**: 1 + max(leftDepth, rightDepth)
- **Depth Calculation**: Accumulate depth as we return from recursion

## 2. PROBLEM CHARACTERISTICS
- **Binary Tree**: Each node has at most 2 children
- **Depth Definition**: Number of nodes from root to deepest leaf
- **Tree Structure**: May be unbalanced, empty, or single node
- **Recursive Nature**: Natural fit for DFS approach

## 3. SIMILAR PROBLEMS
- Minimum Depth of Binary Tree (LeetCode 111)
- Balanced Binary Tree (LeetCode 110)
- Invert Binary Tree (LeetCode 226)
- Same Tree (LeetCode 100)

## 4. KEY OBSERVATIONS
- **Depth propagation**: Each node contributes 1 to depth
- **Maximum selection**: Take max of left and right subtrees
- **Leaf termination**: Depth calculation stops at leaves
- **Linear traversal**: Visit each node exactly once

## 5. VARIATIONS & EXTENSIONS
- **Minimum Depth**: Find shortest path to leaf
- **Balanced Check**: Verify height difference ≤ 1
- **Level Order**: BFS alternative for depth
- **N-ary Tree**: Extend to nodes with multiple children

## 6. INTERVIEW INSIGHTS
- Always clarify: "Is this a binary tree? Can it be empty?"
- Edge cases: empty tree, single node, unbalanced tree
- Space complexity: O(H) where H is tree height (recursion stack)
- Time complexity: O(N) - visit each node once

## 7. COMMON MISTAKES
- Not handling nil root case properly
- Forgetting to add 1 for current node
- Using sum instead of max for depth
- Stack overflow on very deep trees
- Not considering iterative BFS alternative

## 8. OPTIMIZATION STRATEGIES
- **Current solution is optimal**: O(N) time, O(H) space
- **Iterative BFS**: Can use queue to avoid recursion stack
- **Tail recursion**: Not applicable in Go for tree traversal
- **Early pruning**: Not applicable for depth calculation

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like counting the levels in a company hierarchy:**
- You have an organizational chart (tree structure)
- You want to know the longest chain of command
- Start from the CEO (root) and count down to the deepest employee
- For each manager, find their longest reporting chain
- The company depth is 1 + the longest chain among all departments
- Keep track of the maximum depth as you explore the organization

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Binary tree root node
2. **Goal**: Find maximum depth (number of nodes from root to deepest leaf)
3. **Output**: Integer representing maximum depth
4. **Constraint**: Tree may be empty or unbalanced

#### Phase 2: Key Insight Recognition
- **"Recursive structure"** → Tree depth = 1 + max(subtree depths)
- **"Base case"** → Empty tree has depth 0
- **"Divide and conquer"** → Solve left and right subtrees independently
- **"Combine results"** → Take maximum and add current node

#### Phase 3: Strategy Development
```
Human thought process:
"I need to find the deepest level of the tree.
For any node, the depth is 1 (for itself) plus the maximum
of the depths of its left and right subtrees.
If a node is nil, its depth is 0.
I'll recursively calculate depths and combine results."
```

#### Phase 4: Edge Case Handling
- **Empty tree**: Return 0 (no nodes)
- **Single node**: Return 1 (just the root)
- **Unbalanced tree**: Handle different depths naturally
- **Very deep tree**: Consider stack overflow limits

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Tree:     3
         / \
        9   20
           /  \
          15   7

Human thinking:
"I'll calculate depth from bottom up:

Node 9: no children → depth = 1
Node 15: no children → depth = 1
Node 7: no children → depth = 1

Node 20: left depth=1, right depth=1
         max(1,1) + 1 = 2

Node 3: left depth=1, right depth=2
        max(1,2) + 1 = 3

Final answer: 3"
```

#### Phase 6: Intuition Validation
- **Why recursion works**: Tree has natural recursive structure
- **Why max is used**: Want deepest path, not sum of all paths
- **Why O(N) time**: Each node visited exactly once
- **Why O(H) space**: Recursion depth equals tree height

### Common Human Pitfalls & How to Avoid Them
1. **"Why not count edges instead of nodes?"** → Problem specifies nodes, clarify in interview
2. **"Should I use BFS instead?"** → BFS works but uses more memory for wide trees
3. **"What about very deep trees?"** → Consider iterative approach to avoid stack overflow
4. **"Can I optimize further?"** → Already optimal O(N) time

### Real-World Analogy
**Like finding the longest family generation chain:**
- You have a family tree showing parent-child relationships
- You want to know the longest generation line
- Start from the oldest ancestor and trace down
- For each person, find their longest descendant line
- The family depth is 1 + the longest line among all children
- Keep track of the maximum generation count found

### Human-Readable Pseudocode
```
function maxDepth(node):
    if node is nil:
        return 0
    
    leftDepth = maxDepth(node.left)
    rightDepth = maxDepth(node.right)
    
    return max(leftDepth, rightDepth) + 1
```

### Execution Visualization

### Example Tree:
```
     3
    / \
   9  20
     /  \
    15   7
```

### Recursive Call Stack:
```
maxDepth(3)
├── maxDepth(9) → returns 1
├── maxDepth(20)
│   ├── maxDepth(15) → returns 1
│   ├── maxDepth(7) → returns 1
│   └── max(1,1) + 1 = 2
└── max(1,2) + 1 = 3
```

### Key Visualization Points:
- **Base case**: nil nodes return 0
- **Leaf nodes**: Return 1 (no children)
- **Internal nodes**: Return 1 + max of children
- **Root calculation**: Final depth from entire tree

### Memory Layout Visualization:
```
Call Stack (top to bottom):
maxDepth(3) ← waiting for children
maxDepth(20) ← waiting for children  
maxDepth(15) ← returns 1
```

### Time Complexity Breakdown:
- **Each node visited**: O(1) work per node
- **Total nodes**: N
- **Total time**: O(N)
- **Space**: O(H) - maximum recursion depth

### Alternative Approaches:

#### 1. Iterative BFS (O(N) time, O(W) space where W is max width)
```go
func maxDepthBFS(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    queue := []*TreeNode{root}
    depth := 0
    
    for len(queue) > 0 {
        levelSize := len(queue)
        depth++
        
        for i := 0; i < levelSize; i++ {
            node := queue[0]
            queue = queue[1:]
            
            if node.Left != nil {
                queue = append(queue, node.Left)
            }
            if node.Right != nil {
                queue = append(queue, node.Right)
            }
        }
    }
    
    return depth
}
```
- **Pros**: No recursion stack, works for very deep trees
- **Cons**: Uses more memory for wide trees

#### 2. Iterative DFS with Stack (O(N) time, O(H) space)
```go
func maxDepthDFS(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    stack := []struct {
        node  *TreeNode
        depth int
    }{{root, 1}}
    
    maxDepth := 0
    
    for len(stack) > 0 {
        current := stack[len(stack)-1]
        stack = stack[:len(stack)-1]
        
        node := current.node
        depth := current.depth
        
        if node.Left == nil && node.Right == nil {
            maxDepth = max(maxDepth, depth)
        }
        
        if node.Right != nil {
            stack = append(stack, struct {
                node  *TreeNode
                depth int
            }{node.Right, depth + 1})
        }
        
        if node.Left != nil {
            stack = append(stack, struct {
                node  *TreeNode
                depth int
            }{node.Left, depth + 1})
        }
    }
    
    return maxDepth
}
```
- **Pros**: No recursion, same space complexity
- **Cons**: More complex implementation

#### 3. Morris Traversal (O(N) time, O(1) space)
```go
func maxDepthMorris(root *TreeNode) int {
    if root == nil {
        return 0
    }
    
    maxDepth := 0
    current := root
    depth := 0
    
    for current != nil {
        if current.Left == nil {
            current = current.Right
            depth++
        } else {
            predecessor := current.Left
            depthTemp := 1
            
            for predecessor.Right != nil && predecessor.Right != current {
                predecessor = predecessor.Right
                depthTemp++
            }
            
            if predecessor.Right == nil {
                predecessor.Right = current
                current = current.Left
                depth++
            } else {
                predecessor.Right = nil
                current = current.Right
                maxDepth = max(maxDepth, depth)
                depth = depthTemp
            }
        }
    }
    
    return maxDepth
}
```
- **Pros**: O(1) space, no recursion or stack
- **Cons**: Very complex, modifies tree structure

### Extensions for Interviews:
- **Minimum Depth**: Find shortest path to leaf
- **Balanced Tree**: Check if height-balanced
- **N-ary Tree**: Extend to multiple children
- **Maximum Width**: Find widest level of tree
*/
func main() {
	// Test cases
	testCases := [][]interface{}{
		{3, 9, 20, nil, nil, 15, 7},
		{1, nil, 2},
		{},
		{1},
		{1, 2, 3, 4, 5, nil, nil, 6, 7, nil, nil, nil, nil, 8},
		{1, 2, 3, 4, 5, 6, 7},
		{1, 2, 3, 4, nil, nil, nil, 5},
		{1, nil, 2, nil, 3, nil, 4, nil, 5},
		{0},
		{1, 2},
	}
	
	for i, nums := range testCases {
		root := createTree(nums)
		result1 := maxDepth(root)
		result2 := maxDepthBFS(root)
		
		fmt.Printf("Test Case %d: %v\n", i+1, nums)
		fmt.Printf("  Recursive: %d, BFS: %d\n\n", result1, result2)
	}
}

File: long_questions/DSA/Leetcode/java/10_Tree_DFS/104_Maximum_Depth_of_Binary_Tree.java
import java.util.*;

public class MaximumDepthOfBinaryTree {
    
    // Definition for a binary tree node.
    public static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        
        TreeNode() {}
        TreeNode(int val) { this.val = val; }
        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    // 104. Maximum Depth of Binary Tree
    // Time: O(N), Space: O(H) where H is tree height (recursion stack)
    public static int maxDepth(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        int leftDepth = maxDepth(root.left);
        int rightDepth = maxDepth(root.right);
        
        return Math.max(leftDepth, rightDepth) + 1;
    }

    // Iterative BFS approach
    public static int maxDepthBFS(TreeNode root) {
        if (root == null) {
            return 0;
        }
        
        Queue<TreeNode> queue = new LinkedList<>();
        queue.offer(root);
        int depth = 0;
        
        while (!queue.isEmpty()) {
            int levelSize = queue.size();
            depth++;
            
            for (int i = 0; i < levelSize; i++) {
                TreeNode node = queue.poll();
                
                if (node.left != null) {
                    queue.offer(node.left);
                }
                if (node.right != null) {
                    queue.offer(node.right);
                }
            }
        }
        
        return depth;
    }

    // Helper function to create a binary tree from array
    public static TreeNode createTree(Object[] nums) {
        if (nums.length == 0) {
            return null;
        }
        
        TreeNode[] nodes = new TreeNode[nums.length];
        for (int i = 0; i < nums.length; i++) {
            if (nums[i] == null) {
                nodes[i] = null;
            } else {
                nodes[i] = new TreeNode((Integer) nums[i]);
            }
        }
        
        for (int i = 0; i < nums.length; i++) {
            if (nodes[i] != null) {
                int left = 2 * i + 1;
                int right = 2 * i + 2;
                
                if (left < nums.length) {
                    nodes[i].left = nodes[left];
                }
                if (right < nums.length) {
                    nodes[i].right = nodes[right];
                }
            }
        }
        
        return nodes[0];
    }

    public static void main(String[] args) {
        // Test cases
        Object[][] testCases = {
            {3, 9, 20, null, null, 15, 7},
            {1, null, 2},
            {},
            {1},
            {1, 2, 3, 4, 5, null, null, 6, 7, null, null, null, null, 8},
            {1, 2, 3, 4, 5, 6, 7},
            {1, 2, 3, 4, null, null, null, 5},
            {1, null, 2, null, 3, null, 4, null, 5},
            {0},
            {1, 2}
        };
        
        for (int i = 0; i < testCases.length; i++) {
            TreeNode root = createTree(testCases[i]);
            int result1 = maxDepth(root);
            int result2 = maxDepthBFS(root);
            
            System.out.printf("Test Case %d: %s\n", i + 1, Arrays.toString(testCases[i]));
            System.out.printf("  Recursive: %d, BFS: %d\n\n", result1, result2);
        }
    }

/*
=======================================
PATTERN RECOGNITION & INSIGHTS
=======================================

## 1. ALGORITHM PATTERN: Tree Depth-First Search (DFS)
- **Recursive DFS**: Explore tree depth-first using recursion
- **Base Case**: Null node has depth 0
- **Recursive Case**: Depth = 1 + max(leftDepth, rightDepth)
- **Call Stack**: Uses recursion stack for traversal

## 2. PROBLEM CHARACTERISTICS
- **Binary Tree**: Each node has at most 2 children
- **Depth Definition**: Number of nodes from root to leaf
- **Longest Path**: Find maximum depth among all root-to-leaf paths
- **Tree Structure**: Can be balanced or unbalanced

## 3. SIMILAR PROBLEMS
- Minimum Depth of Binary Tree
- Balanced Binary Tree
- Diameter of Binary Tree
- Path Sum problems

## 4. KEY OBSERVATIONS
- Depth is 1 + max(leftDepth, rightDepth)
- Base case: null node has depth 0
- Recursive call explores both subtrees
- Maximum depth comes from the longest root-to-leaf path
- Time complexity: O(N) - each node visited once

## 5. VARIATIONS & EXTENSIONS
- Iterative BFS approach
- Find depth of specific node
- Handle N-ary trees
- Return the actual deepest path

## 6. INTERVIEW INSIGHTS
- Clarify: "Is tree guaranteed to be binary?"
- Edge cases: empty tree, single node, skewed tree
- Time complexity: O(N) vs O(N²) naive approach
- Space complexity: O(H) vs O(N) for iterative

## 7. COMMON MISTAKES
- Not handling null base case
- Off-by-one errors in depth calculation
- Stack overflow for very deep trees
- Confusing depth with node count

## 8. OPTIMIZATION STRATEGIES
- Recursive approach is optimal for single query
- Iterative BFS avoids recursion stack limits
- Tail recursion optimization
- Memoization for multiple queries

## 9. EXECUTION VISUALIZATION

## 10. HUMAN LOGIC PHASE

### Mental Model & Intuition
**Think of it like exploring a family tree:**
- You're at the root (ancestor)
- You want to find the longest generation chain
- You explore each branch (left and right children)
- For each person, you ask about their descendants
- The longest chain of descendants gives you the depth

### Step-by-Step Human Reasoning

#### Phase 1: Problem Understanding
1. **Input**: Root of binary tree
2. **Goal**: Find maximum depth (longest root-to-leaf path)
3. **Output**: Integer representing maximum depth

#### Phase 2: Key Insight Recognition
- **"How to measure depth?"** → Count nodes from root to leaf
- **"What's the base case?"** → Null node has depth 0
- **"How to combine results?"** → 1 + max(left, right)
- **"Can I use iteration?"** → Yes, but recursion is natural

#### Phase 3: Strategy Development
```
Human thought process:
"I'll use recursive DFS:
1. If current node is null, return depth 0
2. Otherwise:
   - Recursively find depth of left subtree
   - Recursively find depth of right subtree
   - Return 1 + maximum of these two depths
3. This naturally finds the longest path"
```

#### Phase 4: Edge Case Handling
- **Empty tree**: Return 0 (no nodes)
- **Single node**: Return 1 (just the root)
- **Skewed tree**: Recursion depth equals number of nodes
- **Balanced tree**: Depth is log₂(N) approximately

#### Phase 5: Algorithm Walkthrough (Human Perspective)
```
Tree:    3
        /   \
       9   20
      / \   / \
     15   7

Human thinking:
"Start at root (3):
- Left subtree depth?
  - Node 9: left is null (depth 0), right is null (depth 0)
  - So depth of node 9 = 1 + max(0,0) = 1
- Right subtree depth?
  - Node 20: left is 15, right is 7
  - Need depth of node 15:
    - Left is null (0), right is null (0)
    - So depth of 15 = 1 + max(0,0) = 1
  - Need depth of node 7:
    - Both children are null (0)
    - So depth of 7 = 1 + max(0,0) = 1
  - So depth of 20 = 1 + max(1,1) = 2
- Final answer: 1 + max(1,2) = 3

The longest path is 3→9→15 or 3→20→7"
```

#### Phase 6: Intuition Validation
- **Why it works**: Each node contributes 1 to path length
- **Why it's efficient**: Each node visited exactly once
- **Why it's correct**: Recursion naturally explores all paths

### Common Human Pitfalls & How to Avoid Them
1. **"Why not count all paths?"** → Too expensive O(N²)
2. **"What about iterative approach?"** → Possible but less natural
3. **"How to handle null?"** → Always check before accessing children
4. **"What about very deep trees?"** → May cause stack overflow

### Real-World Analogy
**Like finding the longest family generation:**
- You're studying a family tree starting from an ancestor
- You want to find the longest chain of generations
- For each person, you explore their descendants
- The longest chain from ancestor to any descendant gives you the answer
- Each person contributes exactly 1 to their generation level

### Human-Readable Pseudocode
```
function maxDepth(root):
    if root is null:
        return 0
    
    leftDepth = maxDepth(root.left)
    rightDepth = maxDepth(root.right)
    
    return 1 + max(leftDepth, rightDepth)
```

### Execution Visualization

### Example: Tree with root 3, left 9, right 20; 20 has left 15, right 7
```
Tree Structure:
        3
       /   \
      9   20
     / \   / \
    15   7

DFS Traversal:
Start at root (3):
→ Depth = ?

Explore left subtree (node 9):
→ 9 has no children, depth = 1

Explore right subtree (node 20):
→ 20 has children 15 and 7
→ 
→ Explore 15:
→ 15 has no children, depth = 2
→ 
→ Explore 7:
→ 7 has no children, depth = 2
→ 
→ Depth of 20 = 1 + max(2,2) = 3
→ 
→ Final answer: 1 + max(1,3) = 4

Wait, let me recalculate:

Depth of 9 = 1 (no children)
Depth of 15 = 2 (no children)
Depth of 7 = 2 (no children)
Depth of 20 = 1 + max(2,2) = 3
Final answer: 1 + max(1,3) = 4

The longest path is 3→20→15 (length 3) ✓
```

### Key Visualization Points:
- **Recursive DFS** explores each subtree completely
- **Base case** handles null nodes correctly
- **Depth calculation** adds 1 for current node
- **Maximum function** finds longest path

### Memory Layout Visualization:
```
Call Stack Evolution:
maxDepth(3)
├─ maxDepth(9) → returns 1
└─ maxDepth(20)
   ├─ maxDepth(15) → returns 2
   └─ maxDepth(7) → returns 2
   └─ returns 1 + max(2,2) = 3
└─ returns 1 + max(1,3) = 4

Tree Paths:
3→9 (length 2)
3→20→15 (length 3) ← Longest!
3→20→7 (length 3)
```

### Time Complexity Breakdown:
- **Each Node**: Visited exactly once
- **Recursive Calls**: 2 calls per node (left and right)
- **Total**: O(N) time where N is number of nodes
- **Space**: O(H) for recursion stack, H is tree height
- **Optimal**: Cannot do better than O(N) for this problem
*/
